src/bsoncxx/string/view_or_value.hpp

```diff
@@ -0,0 +1,77 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <string>
+
+#include <bsoncxx/stdx/string_view.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+#include <iostream>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace string {
+
+using string_variant = bsoncxx::view_or_value<stdx::string_view, std::string>;
+
+///
+/// Class representing a view-or-value variant type for strings.
+///
+class BSONCXX_API view_or_value : public string_variant {
+   public:
+    ///
+    /// Forward all constructors.
+    ///
+    using string_variant::string_variant;
+
+    ///
+    /// Construct a string::view_or_value using a null-terminated const char *.
+    ///
+    /// @param str A null-terminated string
+    ///
+    BSONCXX_INLINE view_or_value(const char* str)
+        : bsoncxx::view_or_value<stdx::string_view, std::string>(std::string(str)) {
+    }
+
+    ///
+    /// Return a null-terminated string from this string::view_or_value.
+    ///
+    /// @note if we are non-owning, we cannot assume our string_view is null-terminated.
+    /// In this case we must make a new, owning view_or_value with a null-terminated
+    /// copy of our view.
+    ///
+    /// @return A pointer to the original string, or a null-terminated copy.
+    ///
+    BSONCXX_INLINE const char* c_str() {
+        // If we own our string, it must be a std::string, return a const char*
+        if (_value) {
```

Comment:
I think this method might be simpler if you reversed the logic:

if (!_value) {
    // become value-y
}

return _value->c_stR()